api/python/tests/integration/test_packages.py

```diff
@@ -1410,3 +1416,70 @@ def test_resolve_hash(self):
+    [('incoming', 6, {'name': 'incoming'}, None), ('existing', 1, {'name': 'existing'}, None)]
+)
+def test_set_dir_update_policy(update_policy, expected_size, expected_meta, expected_hash):
+    """Verify building a package with update policy. """
+
+    data_dir = pathlib.Path(__file__).parent / "data"
+    nested_dir = data_dir / 'nested'
+    nested_dir_2 = data_dir / 'nested2'
+    pkg = Package()
+    pkg = pkg.set_dir("/", nested_dir, meta={'name': 'test_meta'})
+    assert 'one.txt' in pkg.keys()
+    assert pkg.meta == {'name': 'test_meta'}
+
+    pkg = pkg.set_dir("/", nested_dir_2, meta={'name': update_policy}, update_policy=update_policy)
+    assert 'one.txt' in pkg.keys()
+    assert pkg['one.txt'].size == expected_size
+    assert pkg['one.txt'].hash == expected_hash
+    assert pkg.meta == expected_meta
+
+    # verify non existing update policy raises value error
+    expected_err = f"Update policy should be one of {str(PACKAGE_UPDATE_POLICY)}, not 'invalid_policy'"
+    with pytest.raises(ValueError) as e:
+        pkg.set_dir("nested", data_dir, update_policy='invalid_policy')
+    assert expected_err in str(e.value)
+
+
+@pytest.mark.parametrize(
+    'update_policy, expected_a_url, expected_xy_url',
+    [
+        ('existing', 's3://bucket/foo/a.txt?versionId=xyz', 's3://bucket/foo/x/y.txt?versionId=null'),
+        ('incoming', 's3://bucket/bar/a.txt?versionId=abc', 's3://bucket/bar/x/y.txt?versionId=null')
+    ]
+)
+def test_set_dir_update_policy_s3(update_policy, expected_a_url, expected_xy_url):
+    with patch('quilt3.packages.list_object_versions') as list_object_versions_mock:
+        list_object_versions_mock.return_value = (
+            [
+                dict(Key='foo/a.txt', VersionId='xyz', IsLatest=True, Size=10),
+                dict(Key='foo/x/y.txt', VersionId='null', IsLatest=True,
+                     Size=10),
+                dict(Key='foo/z.txt', VersionId='123', IsLatest=False, Size=10),
+            ], [])
+        pkg = Package()
+        pkg.set_dir('', 's3://bucket/foo/', meta={'name': 'test_meta'})
+        assert pkg['a.txt'].get() == 's3://bucket/foo/a.txt?versionId=xyz'
+        assert pkg['x/y.txt'].get() == 's3://bucket/foo/x/y.txt?versionId=null'
+        list_object_versions_mock.assert_called_once_with('bucket', 'foo/')
+
+        list_object_versions_mock.return_value = (
+            [
+                dict(Key='bar/a.txt', VersionId='abc', IsLatest=True, Size=10),
+                dict(Key='bar/x/y.txt', VersionId='null', IsLatest=True,
+                     Size=10),
+                dict(Key='bar/z.txt', VersionId='123', IsLatest=False, Size=10),
+            ], [])
+
+        pkg.set_dir('', 's3://bucket/bar', update_policy=update_policy)
+        assert pkg['a.txt'].get() == expected_a_url
+        assert pkg['x/y.txt'].get() == expected_xy_url
+        calls = [call('bucket', 'foo/'), call('bucket', 'bar/')]
+        list_object_versions_mock.assert_has_calls(calls)
```

Comment:
you can just inline the calls here, easier to read

Reply:
and remember you need to check the call_count as well. this assert will succeed if it gets called 40 times, even though that's a bug.